jerry-core/debugger/debugger.c

```diff
@@ -195,6 +196,305 @@ jerry_debugger_send_backtrace (const uint8_t *recv_buffer_p) /**< pointer to the
+                                        ecma_string_t *value_str, /**< property name or value string */
+                                        jerry_debugger_send_string_t *message_string_p, /**< message pointer */
+                                        size_t *buffer_pos) /**< string data position of the message */
+{
+  const size_t max_byte_count = JERRY_DEBUGGER_SEND_MAX (uint8_t);
+  const size_t max_message_size = JERRY_DEBUGGER_SEND_SIZE (max_byte_count, uint8_t);
+
+  ECMA_STRING_TO_UTF8_STRING (value_str, str_buff, str_buff_size);
+
+  int copied_values = 3; // type, length, value
+  size_t str_size = 0;
+  size_t str_limit = 255;
+
+  while (copied_values > 0)
+  {
+    if (*buffer_pos == max_byte_count)
+    {
+      if (!jerry_debugger_send (max_message_size))
+      {
+        ECMA_FINALIZE_UTF8_STRING (str_buff, str_buff_size);
+        return false;
+      }
+
+      *buffer_pos = 0;
+    }
+
+    if (copied_values == 3)
+    {
+      if (variable_type != JERRY_DEBUGGER_VARIABLE_NONE)
+      {
+        memcpy (message_string_p->string + *buffer_pos, &variable_type, 1);
+        *buffer_pos += 1;
+      }
+    }
+    else if (copied_values == 2)
+    {
+      str_size = (str_buff_size > str_limit) ? str_limit : str_buff_size;
+
+      memcpy (message_string_p->string + *buffer_pos, &str_size, 1);
+      *buffer_pos += 1;
+    }
+    else
+    {
+      size_t free_bytes = max_byte_count - *buffer_pos;
+      const uint8_t *string_p = str_buff;
+
+      while (str_size > free_bytes)
+      {
+        memcpy (message_string_p->string + *buffer_pos, string_p, free_bytes);
+
+        if (!jerry_debugger_send (max_message_size))
+        {
+          ECMA_FINALIZE_UTF8_STRING (str_buff, str_buff_size);
+
+          return false;
+        }
+
+        string_p += free_bytes;
+        str_size -= free_bytes;
+        free_bytes = max_byte_count;
+        *buffer_pos = 0;
+      }
+
+      memcpy (message_string_p->string + *buffer_pos, string_p, str_size);
+      *buffer_pos += str_size;
+    }
+
+    copied_values--;
+  }
+
+  ECMA_FINALIZE_UTF8_STRING (str_buff, str_buff_size);
+
+  return true;
+} /* copy_scope_variables_to_string_message */
+
+/**
+ * Send variables of the given scope chain level.
+ */
+static void
+jerry_debugger_send_scope_variables (const uint8_t *recv_buffer_p) /**< pointer to the received data */
+{
+  JERRY_DEBUGGER_RECEIVE_BUFFER_AS (jerry_debugger_receive_get_scope_variables_t, get_scope_variables_p);
+
+  uint32_t chain_index;
+  memcpy (&chain_index, get_scope_variables_p->chain_index, sizeof (uint32_t));
+
+  vm_frame_ctx_t *iter_frame_ctx_p = JERRY_CONTEXT (vm_top_context_p);
+  ecma_object_t *lex_env_p = iter_frame_ctx_p->lex_env_p;
+
+  while (chain_index != 0)
+  {
+    lex_env_p = ecma_get_lex_env_outer_reference (lex_env_p);
+
+    if (lex_env_p == NULL)
+    {
+      jerry_debugger_send_type (JERRY_DEBUGGER_SCOPE_VARIABLES_END);
+      return;
+    }
+    chain_index--;
```

Comment:
Ignore non this object or declarative lex envs as well.

Reply:
I do not understand exactly what you mean. It will steps through the outer references of the current lexical environment until the `chain_index` arguments reach zero. It is always touch a declarative or this object bound lexical environment. If not, it will returns with an empty message with `JERRY_DEBUGGER_SCOPE_VARIABLES_END` type (e.g.: the `chain_index` argument is invalid / bigger than the scope chain).

Reply:
In the future classes will introduces a new scope type. This type should not be shared by the debugger client side, and when encountered it should be skipped. Therefore only the two currently supported types should be handled by this loops.

Reply:
Oh I see, thanks. I've updated it.